File: Lintcode/167-Add-Two-Numbers/AddTwoNumbers.java

/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */

public class AddTwoNumbers {
	/**
     * @param l1: the first list
     * @param l2: the second list
     * @return: the sum list of l1 and l2
     */
    public ListNode addLists(ListNode l1, ListNode l2) {
        // write your code here
        if (l1 == null) {
            return l2;
        }
        if (l2 == null) {
            return l1;
        }
        int carry = 0;
        ListNode cur1 = l1;
        ListNode cur2 = l2;
        ListNode sumHead = null;
        ListNode curSum = null;
        while (cur1 != null || cur2 != null)  {
            if (cur1 != null) {
                carry += cur1.val;
                cur1 = cur1.next;
            }
            if (cur2 != null) {
                carry += cur2.val;
                cur2 = cur2.next;
            }
            if (sumHead == null) {
                sumHead = new ListNode(carry >= 10 ? carry - 10 : carry);
                curSum = sumHead;
                carry = carry / 10;
                continue;
            }
            curSum.next = new ListNode(carry >= 10 ? carry - 10 : carry);
            curSum = curSum.next;
            carry = carry / 10;
        }
        // Assuming l1 and l2 are of the same length.
        if (carry != 0) {
            curSum.next = new ListNode(carry);
            curSum = curSum.next;
        }
        return sumHead;
    }// end of addLists
}

File: Leetcode/2-Add-Two-Numbers/AddTwoNumbers.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    cur1 = l1;
    cur2 = l2;
    dummy = new ListNode(0);
    cur = dummy;
    carry = 0;
    while (cur1 != undefined || cur2 != undefined) {
        if (cur1 != undefined) {
            carry += cur1.val
            cur1 = cur1.next
        }
        
        if (cur2 != undefined) {
            carry += cur2.val
            cur2 = cur2.next
        }
        
        temp = new ListNode(carry % 10)
        cur.next = temp;
        cur = cur.next;
        carry = Math.floor(carry / 10)
    }
    if (carry != 0) {
        temp = new ListNode(carry);
        cur.next = temp
        cur = cur.next
    }
    return dummy.next
};

